hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java

```diff
@@ -234,27 +224,126 @@ private Pair<Integer, Boolean> submitPurgeKeysRequest(List<DeleteBlockGroupResul
+      }
+      if (expectedPreviousSnapshotId != null) {
+        requestBuilder.setExpectedPreviousSnapshotID(
+            NullableUUID.newBuilder()
+                .setUuid(HddsUtils.toProtobuf(expectedPreviousSnapshotId))
+                .build());
+      }
+
+      currSize = requestBuilder.build().getSerializedSize();
+      remainingRatisLimit -= currSize;
+
+      // 3.1 Add purgeKeys
+      List<String> batchPurgeKeys = new ArrayList<>();
+      int estimatedPurgeKeysSize = 0;
+
+      while (purgeKeyIndex < purgeKeys.size()) {
+        String nextKey = purgeKeys.get(purgeKeyIndex);
+
+        List<String> temp = new ArrayList<>(batchPurgeKeys);
+        temp.add(nextKey);
+
+        OzoneManagerProtocolProtos.DeletedKeys deletedKeys = OzoneManagerProtocolProtos.DeletedKeys.newBuilder()
+            .setVolumeName("").setBucketName("").addAllKeys(temp).build();
+
+        PurgeKeysRequest.Builder tempBuilder = requestBuilder.clone();
+        tempBuilder.clearDeletedKeys().addDeletedKeys(deletedKeys);
+
+        estimatedPurgeKeysSize = tempBuilder.build().getSerializedSize();
+        if (currSize + estimatedPurgeKeysSize > remainingRatisLimit) {
+          batchCapacityReached = true;
+          break;
+        }
+
+        batchPurgeKeys.add(nextKey);
+        purgeKeyIndex++;
       }
-    }
 
-    OzoneManagerProtocolProtos.OMRequest omRequest = OzoneManagerProtocolProtos.OMRequest.newBuilder()
-        .setCmdType(OzoneManagerProtocolProtos.Type.PurgeKeys)
-        .setPurgeKeysRequest(purgeKeysRequest)
-        .setClientId(getClientId().toString())
-        .build();
-
-    // Submit PurgeKeys request to OM. Acquire bootstrap lock when processing deletes for snapshots.
-    try (BootstrapStateHandler.Lock lock = snapTableKey != null ? getBootstrapStateLock().lock() : null) {
-      OzoneManagerProtocolProtos.OMResponse omResponse = submitRequest(omRequest);
-      if (omResponse != null) {
-        purgeSuccess = purgeSuccess && omResponse.getSuccess();
+      // Now actually add batchPurgeKeys
+      if (!batchPurgeKeys.isEmpty()) {
+        OzoneManagerProtocolProtos.DeletedKeys deletedKeys = OzoneManagerProtocolProtos.DeletedKeys.newBuilder()
+            .setVolumeName("").setBucketName("")
+            .addAllKeys(batchPurgeKeys).build();
+        requestBuilder.addDeletedKeys(deletedKeys);
+        currSize = requestBuilder.build().getSerializedSize();
+      }
+
+      remainingRatisLimit -= currSize;
+
+      // 3.2 Add keysToUpdate
+      while (!batchCapacityReached && updateIndex < keysToUpdateList.size()) {
+        OzoneManagerProtocolProtos.SnapshotMoveKeyInfos nextUpdate = keysToUpdateList.get(updateIndex);
+
+        PurgeKeysRequest.Builder tempBuilder = requestBuilder.clone();
+        tempBuilder.addKeysToUpdate(nextUpdate);
+        int estimatedSize = tempBuilder.build().getSerializedSize();
+
+        if (currSize + estimatedSize > remainingRatisLimit) {
```

Comment:
This condition is wrong. 
```suggestion
        if (currSize + estimatedSize > ratisLimit) {
```